modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/source/SourceConnectionManager.java

```diff
@@ -51,8 +57,11 @@
   <T> T getConnection(ConfigurationInstance config) throws ConnectionException {
     ConnectionHandler<Object> connectionHandler = connectionManager.getConnection(config.getValue());
     Object connection = connectionHandler.getConnection();
-    connections.put(new Reference<>(connection), connectionHandler);
-
+    Reference<Object> connReference = new Reference<>(connection);
+    withLock(lock, () -> {
+      ((AtomicInteger) connections.computeIfAbsent(connReference, k -> new Pair(new AtomicInteger(0), connectionHandler))
```

Comment:
computeIfAbsent method has performance issues. Replace this map with a caffeine cache.